### PR TITLE
Adds Create Pull Request menu item

### LINKED ISSUE #2135: "Create Pull Request" menu item
Branch > Create Pull Request

1. If the branch has unpushed commits or is unpublished, push.
1. Open the Pull Request view on GitHub.

### PR DESCRIPTION
**Pull request by iAmWillShepherd**
This PR fixes #2135.

- [x] Add menu item to UI
- [x] Push unpublished commits
- [x] Publish unpublished branch
- [x] Disable menu item if Repository isn't hosted on GitHub
- [ ] ~Open PR view on GitHub and attempt to pre-fill as much info as possible~
- [x] Figure out a decent shortcut

### Unpublished Branch
publish
### Branch Ahead
push

### CODE REVIEW

**Comment by iAmWillShepherd** [comment]
@joshaber should we show a popup letting the user know that they have commits that need to be pushed or a branch that must be publish and allow them to continue? I would prefer this way instead of us automatically doing it. Also, what do you feel is a good short-cut for this?

@donokuda do you have any thoughts?

**Comment by donokuda** [comment]
> @joshaber should we show a popup letting the user know that they have commits that need to be pushed or a branch that must be publish and allow them to continue? I would prefer this way instead of us automatically doing it.

That's a good point. Maybe a dialog along the lines of:

```
+------------------------------------------------+
|          You have 3 unpushed commits           |
+------------------------------------------------+
| Push these commits to `origin/foobar`          |
| before creating your pull request?             |
|------------------------------------------------|
| [Cancel] [ Push commits and open pull request] |
+------------------------------------------------+
```

The only thing I'm not sure of at the moment is that I feel like we should have 2 buttons here (in addition to the cancel):

- Only open pull request
- Push commits and open pull request

However, I'm not sure if we'll be able to fit 3 buttons nicely in the dialog (especially given how wordy the push commits action is.)

For for an unpublished branch we should only have one main action:

- Publish branch and open pull request

**Comment by iAmWillShepherd** [comment]
I feel like we should require commits be pushed here. It's a bit opinionated, but I think it's best for most users. Those who are more advanced still have access to the web UI. I will cede my opinion to you on this however.

**Comment by donokuda** [comment]
> I feel like we should require commits be pushed here. It's a bit opinionated, but I think it's best for most users. Those who are more advanced still have access to the web UI. I will cede my opinion to you on this however.

👍 Totally fair and alleviates my concern! 😄

File: app/src/main-process/menu/build-default-menu.ts [hunk 1]
```diff
@@ -272,6 +272,11 @@ export function buildDefaultMenu(sharedProcess: SharedProcess): Electron.Menu {
         accelerator: 'CmdOrCtrl+Shift+C',
         click: emit('compare-branch'),
       },
+      {
+        label: __DARWIN__ ? 'Create Pull Request' : 'Create Pull Request',
```
Comment (joshaber) [inline app/src/main-process/menu/build-default-menu.ts]: Should this be Sentence case on Windows?

File: app/src/ui/app.tsx [hunk 2]
```diff
@@ -1190,6 +1216,61 @@ export class App extends React.Component<IAppProps, IAppState> {
...
+      return
+    }
+
+    const dispatcher = this.props.dispatcher
+    const repository = selection.repository
+    const branch = tip.branch
+    const aheadBehind = selection.state.aheadBehind
+
+    if (!aheadBehind) {
+      //Pubilish branch
```
Comment (joshaber) [inline app/src/ui/app.tsx]: I'm not sure this comment and the one in the next branch are really adding much.

File: app/src/ui/app.tsx [hunk 3]
```diff
@@ -1190,6 +1216,61 @@ export class App extends React.Component<IAppProps, IAppState> {
...
+        repository,
+        branch,
+        unPushedCommits: aheadBehind.ahead,
+      })
+    } else {
+      this.OpenPullRequestOnGithub(repository, branch)
+    }
+  }
+
+  private OpenPullRequestOnGithub = (
```
Comment (joshaber) [inline app/src/ui/app.tsx]: Style: this ain't C#, methods are lowercased 😛

File: app/src/ui/app.tsx [hunk 4]
```diff
@@ -1190,6 +1216,61 @@ export class App extends React.Component<IAppProps, IAppState> {
...
+  ) => {
+    const gitHubRepository = repository.gitHubRepository
+
+    if (!gitHubRepository) {
+      return
+    }
+
+    const baseURL = `${gitHubRepository.htmlURL}/compare/${gitHubRepository.defaultBranch}...${branch.nameWithoutRemote}`
+
+    if (baseURL) {
```
Comment (joshaber) [inline app/src/ui/app.tsx]: This check seems unnecessary?

File: app/src/ui/app.tsx [hunk 5]
```diff
@@ -1190,6 +1216,61 @@ export class App extends React.Component<IAppProps, IAppState> {
...
+    repository: Repository,
+    branch: Branch
+  ) => {
+    const gitHubRepository = repository.gitHubRepository
+
+    if (!gitHubRepository) {
+      return
+    }
+
+    const baseURL = `${gitHubRepository.htmlURL}/compare/${gitHubRepository.defaultBranch}...${branch.nameWithoutRemote}`
```
Comment (joshaber) [inline app/src/ui/app.tsx]: I think we should be linking to, e.g., `https://github.com/desktop/desktop/pull/new/$BRANCH_NAME`

**Comment by joshaber** [comment]
>should we show a popup letting the user know that they have commits that need to be pushed or a branch that must be publish and allow them to continue? I would prefer this way instead of us automatically doing it.

What are the situations in which the user wants to create a PR but doesn't want to publish their changes?

In the Classic apps, creating a PR always pushed... I'd lean towards that still.

**Comment by joshaber** [comment]
>Also, what do you feel is a good short-cut for this?

I wonder if we should repurpose command-shift-P. It's kind of a bummer to change the meaning of an existing shortcut, but it's much easier to change shortcuts now than it will be once we're 1.0.

File: app/src/main-process/menu/build-default-menu.ts (hunk 1 above)
Comment (iAmWillShepherd) [inline app/src/main-process/menu/build-default-menu.ts]: I'm not sure. I kinda thought Pull Request was like a common noun that we came up with. I don't know tbh.

File: app/src/ui/app.tsx (hunk 2 above)
Comment (iAmWillShepherd) [inline app/src/ui/app.tsx]: They don't, it was for me. Deleted.

File: app/src/ui/app.tsx (hunk 3 above)
Comment (iAmWillShepherd) [inline app/src/ui/app.tsx]: I wish it was! I like to know the types of my functions before their names 🤣

File: app/src/ui/app.tsx (hunk 4 above)
Comment (iAmWillShepherd) [inline app/src/ui/app.tsx]: At the time it was, now I need to make sure ``${gitHubRepository.htmlURL}` is not null or empty or undefined or 0 or w/e JS sees as false.

**Comment by iAmWillShepherd** [comment]
> What are the situations in which the user wants to create a PR but doesn't want to publish their changes? Are those situations already covered by the "Compare on GitHub" menu item?

I think the situation occurs with new users who may not understand it or people like me who don't realize they are about to do something unwise... Notifying a user helps with that pit-of-success thing people sometimes talk about.

**Comment by iAmWillShepherd** [comment]
⏺

**Comment by iAmWillShepherd** [comment]
I feel like I should include some GIF action of these interactions in case someone documents this stuff in the future.

**Comment by iAmWillShepherd** [comment]
💯

File: app/src/ui/branches/PublishBranch.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+          <p>
+            Your branch must be published before opening a pull request. Would
+            you like to publish <b>{this.props.branch.name}</b> and oepn a pull
+            request?
+          </p>
+        </DialogContent>
+
+        <DialogFooter>
+          <ButtonGroup>
+            <Button type="submit">Cancel</Button>
```
Comment (joshaber) [inline app/src/ui/branches/PublishBranch.tsx]: I think this button order is off. The Cancel button should be on the left and the `ButtonGroup` will need the `destructive` prop.

**Comment by joshaber** [comment]
>I think the situation occurs with new users who may not understand it or people like me who don't realize they are about to do something unwise

Yeah, I could see that as part of the "saving face" thing. We can roll with it and see how it goes.

File: app/src/ui/branches/PublishBranch.tsx [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+interface IPublishBranchProps {
+  readonly dispatcher: Dispatcher
+  readonly repository: Repository
+  readonly branch: Branch
+  readonly onPushed: (repository: Repository, branch: Branch) => void
+  readonly onDismissed: () => void
+}
+
+export class PublishBranch extends React.Component<IPublishBranchProps> {
```
Comment (iAmWillShepherd) [inline app/src/ui/branches/PublishBranch.tsx]: That was the initial route I was going to take, but I figured I'd get called out on it, so I made two.

File: app/src/lib/menu-update.ts [hunk 8]
```diff
@@ -181,13 +182,15 @@ function getMenuState(state: IAppState): Map<MenuIDs, IMenuItemState> {
...
     menuStateBuilder.disable('view-repository-on-github')
+    menuStateBuilder.disable('create-pull-request')
 
     if (
       selectedState &&
       selectedState.type === SelectionType.MissingRepository
     ) {
       if (selectedState.repository.gitHubRepository) {
         menuStateBuilder.enable('view-repository-on-github')
+        menuStateBuilder.enable('create-pull-request')
```
Comment (iAmWillShepherd) [inline app/src/lib/menu-update.ts]: Does that mean `menuStateBuilder.enable('view-repository-on-github')` is wrong too?

**Comment by dwilt** [comment]
@iAmWillShepherd Thanks for putting this together man!!!

File: app/src/ui/branches/PublishBranch.tsx (hunk 7 above)
Comment (joshaber) [inline app/src/ui/branches/PublishBranch.tsx]: 😄 Let's do it 👍

Comment (iAmWillShepherd) [inline app/src/ui/branches/PublishBranch.tsx]: Done ❗️

**Comment by iAmWillShepherd** [comment]
@joshaber are we gonna move forward with `CMD+Shift+P` for the shortcut?

**Comment by iAmWillShepherd** [comment]
💯

**Comment by iAmWillShepherd** [comment]
💯

File: app/src/ui/branches/PushBranchCommits.tsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+  readonly onConfirm: (repository: Repository, branch: Branch) => void
+  readonly onDismissed: () => void
+
+  /**
+   * When set to true, the publish branch dialog is returned.
+   *
+   * @type {boolean}
+   * @memberof IPushBranchCommitsProps
+   */
+  readonly publish?: boolean
```
Comment (joshaber) [inline app/src/ui/branches/PushBranchCommits.tsx]: It seems like this and `unPushedCommits` are communicating the same information, so we should be able to infer this from whether `unPushedCommits` is defined/greater than zero.

File: app/src/ui/branches/PushBranchCommits.tsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+   * Used to show the number of commits a branch is ahead by.
+   * This only needs to be set if `publish` is false or undefined.
+   *
+   * @type {number}
+   * @memberof IPushBranchCommitsProps
+   */
+  readonly unPushedCommits?: number
+}
+
+export class PushBranchCommits extends React.Component<
```
Comment (joshaber) [inline app/src/ui/branches/PushBranchCommits.tsx]: Could you add some brief docs for what this component is/does?

**Comment by iAmWillShepherd** [comment]
💯

**Pull request merged**